tls/s2n_x509_validator.c

```diff
@@ -0,0 +1,415 @@
+ * if none of them return true, the cert is considered invalid.
+ */
+static uint8_t verify_host_information(struct s2n_x509_validator *validator, struct s2n_connection *conn, X509 *public_cert) {
+    uint8_t verified = 0;
+
+    /* Check SubjectAltNames before CommonName as per RFC 6125 6.4.4 */
+    STACK_OF(GENERAL_NAME) *names_list = X509_get_ext_d2i(public_cert, NID_subject_alt_name, NULL, NULL);
+    GENERAL_NAME *current_name = NULL;
+    while (!verified && names_list && (current_name = sk_GENERAL_NAME_pop(names_list))) {
+        const char *name = (const char *) M_ASN1_STRING_data(current_name->d.ia5);
+        size_t name_len = (size_t) M_ASN1_STRING_length(current_name->d.ia5);
+
+        verified = conn->verify_host_fn(name, name_len, conn->data_for_verify_host);
+    }
+
+    GENERAL_NAMES_free(names_list);
+
+    /* if none of those were valid, go to the common name. */
+    if (!verified) {
+        X509_NAME *subject_name = X509_get_subject_name(public_cert);
+        if (subject_name) {
+            int next_idx = 0, curr_idx = -1;
+            while ((next_idx = X509_NAME_get_index_by_NID(subject_name, NID_commonName, curr_idx)) >= 0) {
+                curr_idx = next_idx;
+            }
+
+            if (curr_idx >= 0) {
+                ASN1_STRING *common_name =
+                        X509_NAME_ENTRY_get_data(X509_NAME_get_entry(subject_name, curr_idx));
+
+                if (common_name) {
+                    char peer_cn[255];
+                    static size_t peer_cn_size = sizeof(peer_cn);
+                    memset_check(&peer_cn, 0, peer_cn_size);
+                    if (ASN1_STRING_type(common_name) == V_ASN1_UTF8STRING) {
+                        size_t len = (size_t) ASN1_STRING_length(common_name);
+
+                        lte_check(len, sizeof(peer_cn) - 1);
+                        memcpy_check(peer_cn, ASN1_STRING_data(common_name), len);
+                        verified = conn->verify_host_fn(peer_cn, len, conn->data_for_verify_host);
+                    }
+                }
+            }
+        }
+    }
+
+    return verified;
+}
+
+s2n_cert_validation_code
+s2n_x509_validator_validate_cert_chain(struct s2n_x509_validator *validator, struct s2n_connection *conn, uint8_t *cert_chain_in,
+                                       uint32_t cert_chain_len, struct s2n_cert_public_key *public_key_out) {
+
+    if (validator->validate_certificates && !s2n_x509_trust_store_has_certs(validator->trust_store)) {
+        return S2N_CERT_ERR_UNTRUSTED;
+    }
+
+    X509_STORE_CTX *ctx = NULL;
+
+    struct s2n_blob cert_chain_blob = {.data = cert_chain_in, .size = cert_chain_len};
+    struct s2n_stuffer cert_chain_in_stuffer;
+    if (s2n_stuffer_init(&cert_chain_in_stuffer, &cert_chain_blob) < 0) {
+        return S2N_CERT_ERR_INVALID;
+    }
+    if (s2n_stuffer_write(&cert_chain_in_stuffer, &cert_chain_blob) < 0) {
+        return S2N_CERT_ERR_INVALID;
+    }
+
+    uint32_t certificate_count = 0;
```

Comment:
I don't see a max depth in openssl docs anywhere. Adding it as an option in a later iteration seems more desirable to me. Thoughts?

Reply:
A few implementations will default to a chain length of 3 or 4 and/or let you override that value. I think it is a safer option to have a sane max length to prevent someone from bombarding you with a really long cert chain. I would advocate for having a default max now and adding a configurable value later.